pkg/paho-mqtt/contrib/riot_iface.c

```diff
@@ -0,0 +1,217 @@
+        for (int i=0; i<rc; i++) {
+            DEBUG("0x%02X ", buf[i]);
+        }
+        DEBUG("\n");
+    }
+#endif
+#endif
+
+    return rc;
+}
+
+static int mqtt_write(struct Network *n, unsigned char *buf, int len,
+                      int timeout_ms)
+{
+    /* timeout is controlled by upper layer in PAHO */
+    (void) timeout_ms;
+
+    return sock_tcp_write(&n->sock, buf, len);
+}
+
+void NetworkInit(Network *n)
+{
+#ifdef MODULE_LWIP
+    tsrb_init(&tsrb_lwip_tcp, buffer, TSRB_MAX_SIZE);
+#endif
+    n->mqttread = mqtt_read;
+    n->mqttwrite = mqtt_write;
+}
+
+int NetworkConnect(Network *n, char *addr_ip, int port)
+{
+    sock_tcp_ep_t remote = SOCK_IPV6_EP_ANY;
+    char _local_ip[25];
+
+    strncpy(_local_ip, addr_ip, sizeof(_local_ip));
+    remote.family = AF_INET6;
+    if (ipv6_addr_from_str((ipv6_addr_t *)&remote.addr, _local_ip) == NULL) {
+        LOG_WARNING("Error IPv6: unable to parse destination address, trying IPv4\n");
+        /* ipv6_addr_from_str modifies input buffer */
+        strncpy(_local_ip, addr_ip, sizeof(_local_ip));
+        if (ipv4_addr_from_str((ipv4_addr_t *)&remote.addr,
+                               _local_ip) == NULL) {
+            LOG_ERROR("Error IPv4: unable to parse destination address\n");
+            return -1;
+        }
+        remote.family = AF_INET;
+    }
+    remote.port = port;
+
+    int ret = sock_tcp_connect(&n->sock, &remote, 0, 0);
+    if (ret < 0) {
+        LOG_ERROR("paho-mqtt: unable to connect (%d)\n", ret);
+    } else {
+        ret = 0;
+    }
+    return ret;
+}
+
+void NetworkDisconnect(Network *n)
+{
+    sock_tcp_disconnect(&n->sock);
+}
+
+void TimerInit(Timer *timer)
+{
+    timer->set_ticks.ticks64 = 0;
+    timer->ticks_timeout.ticks64 = 0;
+}
+
+char TimerIsExpired(Timer *timer)
+{
+    return (TimerLeftMS(timer) == 0);
+}
+
+void TimerCountdownMS(Timer *timer, unsigned int timeout_ms)
+{
+    timer->set_ticks = xtimer_now64();
+    timer->ticks_timeout = xtimer_ticks_from_usec64(timeout_ms * US_PER_MS);
+}
+
+void TimerCountdown(Timer *timer, unsigned int timeout_s)
+{
+    TimerCountdownMS(timer, timeout_s * MS_PER_SEC);
+}
+
+int TimerLeftMS(Timer *timer)
+{
+    xtimer_ticks64_t diff_ticks = xtimer_diff64(xtimer_now64(), timer->set_ticks);  /* should be always greater than 0 */
+    if (xtimer_less64(diff_ticks ,timer->ticks_timeout)) {
+        return (xtimer_usec_from_ticks64(xtimer_diff64(timer->ticks_timeout, diff_ticks)) / US_PER_MS);
+    }
+    return 0;
+}
+
+void MutexInit(Mutex *mutex)
+{
+    mutex_init(&mutex->lock);
+}
+
+int MutexLock(Mutex *mutex)
+{
+    mutex_lock(&mutex->lock);
+    return 0;
+}
+
+int MutexUnlock(Mutex *mutex)
+{
+    mutex_unlock(&mutex->lock);
+    return 0;
+}
+
+void *mqtt_riot_run(void *arg)
+{
+    int rc;
+    while (1) {
+        if (arg) {
+            MQTTClient *client = (MQTTClient *)arg;
+
+            MutexLock(&client->mutex);
+            if ((rc = MQTTYield(client, 500)) != 0) {
```

Comment:
Better define a constant with a meaningful name, also containing the unit, than a bare hardcoded value of 500.